### ISSUE TITLE
DataParallelTable and BatchNormalization Bug

### ISSUE THREAD

**Issue by LukeJaffe** [torch/cunn#367]
There is some very strange behavior going on when a DataParallelTable with a BatchNormalization layer is cloned, in which the running_mean and running_var are not properly copied/used by the clone.

We know that data-level parallelization (with DataParallelTable) does not perfectly mirror the standard serial approach for batch normalization, because mini-batches are divided n-fold by the number of GPU devices used. Mean and variance will (probably) not be the same between these sub-mini-batches, since they contain different elements. Nonetheless, when a given model is cloned (DPT or not), the clone should produce exactly the same output as the original (assuming all components are deterministic).

How are running_mean and running_var stored in the torch7 BatchNormalization layer? And specifically, how does DataParallelTable access them? Since running_mean and running_var are not learnable parameters, they are not handled in the standard ways e.g. via model:getParameters(). Does this mean that there is a single copy of running_mean and running_var stored for the DPT? If so, how are they updated? And in this case, how are they accessed, and how do they behave after cloning?

I have attached a very minimal script to demonstrate this issue. This script does a single pass of a small input through a basic DPT model (which contains just Reshape and BatchNormalization layers). After one call of model:forward, the running_mean and running_var of the layer will be updated. Then, we clone the model, and put both the clone and original in evaluate mode. We can see that the output of these two models is not the same. In particular, the original produces a correct output, and the clone produces an anomalous one (not sure how it is calculated). When using 2 GPUs, exactly half of the outputs match between original and clone. You can scale the number of inputs to see this is true.

Further, when using n GPUs, exactly total/n of the outputs match between the original and clone. This suggests that the running_mean and running_var for the first device in the DPT are stored/accessed correctly, but the running_mean, running_var for the others are not.

Testing my script with the given inputs, the last line of the output should print:
Num matching elements: 2/4

I have replicated this result on 3 different machines, with various combinations of NVIDIA drivers and CUDA versions.

[example.zip](https://github.com/torch/cunn/files/576393/example.zip)

**Comment by colesbury** [comment]
The `running_mean` and `running_var` are computed per-replica and never synchronized, except when you construct the DataParallelTable. When you serialized the DataParallelTable, only the first replica is saved, so the `running_mean` and `running_var` from the other replicas are discarded.

For training, this doesn't matter: running_mean and running_var are computed during training, but only used during validation. If everything is deterministic, the output and gradients from training with DPT are equivalent to splitting the batch and accumulating the gradients.

For validation, it's a little trickier: if you create your DPT, train for a bit, and then run validation on the same DPT you can get different results than if you had serialized your DPT between training and validation (`:clone()` serializes)

**Comment by LukeJaffe** [comment]
Is there a way to save all the replicas from the DPT? I have found that when I train a large network with several BatchNormalization layers, the difference in validation performance between the original and serialized DPT is very large.
